File: src/main/solved/Programmers/level0/java/옹알이 (1).java
import java.util.*;

class Solution {
    private Set<String> availableWordSet = new HashSet<>();
    private String[] words = {"aya", "ye", "woo", "ma"};

    public int solution(String[] babbling) {
        int answer = 0;

        makeWords(0, "", new boolean[4]);

        for (String word : babbling) {
            if (availableWordSet.contains(word)) {
                answer += 1;
            }
        }
        return answer;
    }

    private void makeWords(int start, String word, boolean[] visited) {
        if (start == 4) {
            availableWordSet.add(word);
            return ;
        }

        for (int i=0; i<4; i++) {
            if (!visited[i]) {
                visited[i] = true;
                availableWordSet.add(word);
                makeWords(start+1, word+words[i], visited);
                visited[i] = false;
            }
        }
    }
}

File: src/main/solved/Programmers/level0/kotlin/옹알이 (1).kt
import java.util.*

class Solution {
    private val wordSet = HashSet<String>()
    private val words = arrayOf("aya", "ye", "woo", "ma")

    fun solution(babbling: Array<String>): Int {
        var answer: Int = 0

        makeCase(0, "", Array(5){i -> false})

        for (b in babbling) {
            if (wordSet.contains(b)) {
                answer += 1
            }
        }
        return answer
    }

    private fun makeCase(start: Int, word: String, visited: Array<Boolean>) {
        if (start == 4) {
            wordSet.add(word)
            return
        }

        for (i in 0 until 4) {
            if (!visited[i]) {
                visited[i] = true
                wordSet.add(word)
                makeCase(start+1, word+words[i], visited)
                visited[i] = false
            }
        }

    }
}